kernel/examples/read-table-single-threaded/src/main.rs

```diff
@@ -113,26 +114,31 @@ fn try_main() -> DeltaResult<()> {
         .with_schema_opt(read_schema_opt)
         .build()?;
 
-    let mut batches = vec![];
-    for res in scan.execute(engine.as_ref())?.into_iter() {
-        let data = res.raw_data?;
-        let record_batch: RecordBatch = data
-            .into_any()
-            .downcast::<ArrowEngineData>()
-            .map_err(|_| delta_kernel::Error::EngineDataType("ArrowEngineData".to_string()))?
-            .into();
-        let batch = if let Some(mut mask) = res.mask {
-            let extra_rows = record_batch.num_rows() - mask.len();
-            if extra_rows > 0 {
-                // we need to extend the mask here in case it's too short
-                mask.extend(std::iter::repeat(true).take(extra_rows));
-            }
-            filter_record_batch(&record_batch, &mask.into())?
-        } else {
-            record_batch
-        };
-        batches.push(batch);
-    }
+    let batches: Vec<RecordBatch> = scan
+        .execute(engine.as_ref())?
+        .map(|sr_res| {
+            sr_res.and_then(|sr| {
+                let data = sr.raw_data?;
+                let record_batch: RecordBatch = data
+                    .into_any()
+                    .downcast::<ArrowEngineData>()
+                    .map_err(|_| {
+                        delta_kernel::Error::EngineDataType("ArrowEngineData".to_string())
+                    })?
+                    .into();
+                if let Some(mut mask) = sr.mask {
+                    let extra_rows = record_batch.num_rows() - mask.len();
+                    if extra_rows > 0 {
+                        // we need to extend the mask here in case it's too short
+                        mask.extend(std::iter::repeat(true).take(extra_rows));
+                    }
+                    Ok(filter_record_batch(&record_batch, &mask.into())?)
+                } else {
+                    Ok(record_batch)
+                }
+            })
+        })
+        .try_collect()?;
```

Comment:
I find this way which doesn't need a double indent a little cleaner. The `try_collect` will still propagate the error.

```suggestion
let batches: Vec<RecordBatch> = scan
        .execute(engine.as_ref())?
        .map(|sr_res| -> DeltaResult<_> {
            let sr = sr_res?;
            let data = sr.raw_data?;
            let record_batch: RecordBatch = data
                .into_any()
                .downcast::<ArrowEngineData>()
                .map_err(|_| {
                    delta_kernel::Error::EngineDataType("ArrowEngineData".to_string())
                })?
                .into();
            if let Some(mut mask) = sr.mask {
                let extra_rows = record_batch.num_rows() - mask.len();
                if extra_rows > 0 {
                    // we need to extend the mask here in case it's too short
                    mask.extend(std::iter::repeat(true).take(extra_rows));
                }
                Ok(filter_record_batch(&record_batch, &mask.into())?)
            } else {
                Ok(record_batch)
            }
        })
        .try_collect()?;
```